gluten-core/src/main/scala/io/glutenproject/execution/WindowExecTransformer.scala

```diff
@@ -182,6 +172,206 @@ case class WindowExecTransformer(
     }
   }
 
+  private def getWindowRelWithPreProjection(
+      context: SubstraitContext,
+      originalInputAttributes: Seq[Attribute],
+      operatorId: Long,
+      input: RelNode,
+      validation: Boolean): RelNode = {
+    val args = context.registeredFunction
+
+    val preExpressions = new ArrayBuffer[Expression]()
+    val selections = new ArrayBuffer[Int]()
+
+    // Window requires all output attributes from child.
+    preExpressions ++= originalInputAttributes
+    selections ++= originalInputAttributes.indices
+
+    def appendIfNotFound(expression: Expression): Unit = {
+      val foundExpr = preExpressions.find(e => e.semanticEquals(expression))
+      if (foundExpr.isDefined) {
+        // If found, no need to add it to preExpressions again.
+        // The selecting index will be found.
+        selections += preExpressions.indexOf(foundExpr.get)
+      } else {
+        // If not found, add this expression into preExpressions.
+        // A new selecting index will be created.
+        preExpressions += expression.clone()
+        selections += (preExpressions.size - 1)
+      }
+    }
+
+    partitionSpec.foreach(appendIfNotFound)
+    orderSpec.foreach(order => appendIfNotFound(order.child))
+
+    // Create the expression nodes needed by Project node.
+    val preExprNodes = preExpressions
+      .map(
+        ExpressionConverter
+          .replaceWithExpressionTransformer(_, originalInputAttributes)
+          .doTransform(args))
+      .asJava
+    val emitStartIndex = originalInputAttributes.size
+    val inputRel = if (!validation) {
+      RelBuilder.makeProjectRel(input, preExprNodes, context, operatorId, emitStartIndex)
+    } else {
+      // Use a extension node to send the input types through Substrait plan for a validation.
+      val inputTypeNodeList = originalInputAttributes
+        .map(attr => ConverterUtils.getTypeNode(attr.dataType, attr.nullable))
+        .asJava
+      val extensionNode = ExtensionBuilder.makeAdvancedExtension(
+        BackendsApiManager.getTransformerApiInstance.packPBMessage(
+          TypeBuilder.makeStruct(false, inputTypeNodeList).toProtobuf))
+      RelBuilder.makeProjectRel(
+        input,
+        preExprNodes,
+        extensionNode,
+        context,
+        operatorId,
+        emitStartIndex)
+    }
+    getWindowRelAfterProject(
+      context,
+      originalInputAttributes,
+      preExprNodes,
+      selections,
+      inputRel,
+      operatorId,
+      validation)
+  }
+
+  private def getWindowRelAfterProject(
```

Comment:
The name `getWindowRelAfterProject` seems to be easily mixed up with `getWindowRelWithPreProjection`. Could we use a more precise name?

Reply:
Similar to agg, naming functions is hard, just simply modified to 'create'. This part will be removed after inserting pre/post-project through Rule.